File: packages/shared/src/agent/tools/android.js
'use strict';

const fs = require('fs');
const path = require('path');
const { execTool, resolveWorkdir, defineTool } = require('./exec');

/**
 * Android tool — build an Android project through its Gradle wrapper. Delegates
 * to `./gradlew`; no Android build logic is re-implemented. Requires a local
 * JDK + Android SDK (ANDROID_HOME/ANDROID_SDK_ROOT), which the wrapper resolves.
 */

const GRADLE_TASKS = Object.freeze({
  debug: 'assembleDebug',
  release: 'assembleRelease',
  bundle: 'bundleRelease',
  lint: 'lint',
  test: 'testDebugUnitTest',
});

const gradlewCmd = () => (process.platform === 'win32' ? 'gradlew.bat' : './gradlew');

const androidBuildTool = defineTool(
  {
    name: 'android_build',
    description:
      'Build an Android project via its Gradle wrapper. `variant` maps to a Gradle task: debug→assembleDebug, ' +
      'release→assembleRelease, bundle→bundleRelease, lint→lint, test→testDebugUnitTest. Requires a JDK + Android SDK.',
    schema: (z) =>
      z.object({
        variant: z.enum(['debug', 'release', 'bundle', 'lint', 'test']).optional().describe('build variant (default: debug)'),
        module: z.string().optional().describe('Gradle module to scope the task to, e.g. "app"'),
        dir: z.string().optional().describe('workspace-relative project directory (must contain gradlew)'),
      }),
  },
  async (input, ctx) => {
    const dir = resolveWorkdir(ctx, input.dir);
    if (!fs.existsSync(path.join(dir, 'gradlew')) && process.platform !== 'win32') {
      return `❌ android_build: no Gradle wrapper (gradlew) in ${dir}. Run from the Android project root.`;
    }
    const task = GRADLE_TASKS[input.variant || 'debug'];
    const gradleTask = input.module ? `:${sanitizeModule(input.module)}:${task}` : task;
    return execTool({
      ctx,
      label: `android ${input.variant || 'debug'}`,
      command: gradlewCmd(),
      args: [gradleTask, '--stacktrace'],
      dir: input.dir,
      notFoundHint: 'Ensure the Gradle wrapper is present and a JDK + Android SDK (ANDROID_HOME) are installed.',
    });
  }
);

function sanitizeModule(name) {
  const v = String(name || '').trim().replace(/^:/, '');
  if (!/^[A-Za-z0-9_][A-Za-z0-9_.-]*$/.test(v)) throw new Error(`invalid Gradle module: "${name}"`);
  return v;
}

const FACTORIES = Object.freeze({ android_build: androidBuildTool });

module.exports = { FACTORIES, GRADLE_TASKS, sanitizeModule };

File: ai_fleet/agent/tools/android.py
"""Android tool (port of agent/tools/android.js).

Build an Android project through its Gradle wrapper. Delegates to ``./gradlew``;
no Android build logic is re-implemented. Requires a local JDK + Android SDK
(ANDROID_HOME/ANDROID_SDK_ROOT), which the wrapper resolves.
"""

from __future__ import annotations

import os
import re
import sys
from typing import Literal, Optional

from pydantic import BaseModel, Field

from ai_fleet.agent.tools.exec import define_tool, exec_tool, resolve_workdir

GRADLE_TASKS = {
    "debug": "assembleDebug",
    "release": "assembleRelease",
    "bundle": "bundleRelease",
    "lint": "lint",
    "test": "testDebugUnitTest",
}


def gradlew_cmd():
    return "gradlew.bat" if sys.platform.startswith("win") else "./gradlew"


def sanitize_module(name):
    v = re.sub(r"^:", "", str(name or "").strip())
    if not re.match(r"^[A-Za-z0-9_][A-Za-z0-9_.-]*$", v):
        raise Exception(f'invalid Gradle module: "{name}"')
    return v


class _AndroidBuildSchema(BaseModel):
    variant: Optional[Literal["debug", "release", "bundle", "lint", "test"]] = Field(
        default=None, description="build variant (default: debug)"
    )
    module: Optional[str] = Field(default=None, description='Gradle module to scope the task to, e.g. "app"')
    dir: Optional[str] = Field(default=None, description="workspace-relative project directory (must contain gradlew)")


async def _android_build(input, ctx):
    dir_ = resolve_workdir(ctx, input.get("dir"))
    if not os.path.exists(os.path.join(dir_, "gradlew")) and not sys.platform.startswith("win"):
        return f"❌ android_build: no Gradle wrapper (gradlew) in {dir_}. Run from the Android project root."
    variant = input.get("variant") or "debug"
    task = GRADLE_TASKS[variant]
    gradle_task = f":{sanitize_module(input['module'])}:{task}" if input.get("module") else task
    return await exec_tool(
        ctx=ctx,
        label=f"android {variant}",
        command=gradlew_cmd(),
        args=[gradle_task, "--stacktrace"],
        dir=input.get("dir"),
        not_found_hint="Ensure the Gradle wrapper is present and a JDK + Android SDK (ANDROID_HOME) are installed.",
    )


android_build_tool = define_tool(
    {
        "name": "android_build",
        "description": (
            "Build an Android project via its Gradle wrapper. `variant` maps to a Gradle task: debug→assembleDebug, "
            "release→assembleRelease, bundle→bundleRelease, lint→lint, test→testDebugUnitTest. Requires a JDK + Android SDK."
        ),
        "schema": _AndroidBuildSchema,
    },
    _android_build,
)

FACTORIES = {"android_build": android_build_tool}
